Replace All Mappings block

elatoskinas opened the pull request:
# Replace All Mappings
Resolves #2029 by introduced the new replace all mappings block that applies all specified mappings to the given text with the given order.

## Block
![image](https://user-images.githubusercontent.com/34982762/76107255-9411f180-5fd8-11ea-9745-43a56a1ae6ee.png)

`mappings` accept dictionary blocks (and list as a result)
The order specifies which order to replace mappings in, in the case of one mapping being a substring of another.

## Ordering
* **Longest string first:** replace longest first, then replace the shorter substrings in order of string length
* **Dictionary order:** replace by element order in dictionary
* **Earliest occurrence:** replace by first index of occurrence of a substring. If the indices are the same, replace by longest first.

## Example
![blocks](https://user-images.githubusercontent.com/34982762/76107795-b22c2180-5fd9-11ea-874d-fc2e3a5f2847.png)

![image](https://user-images.githubusercontent.com/34982762/76107979-03d4ac00-5fda-11ea-8699-3f2746651425.png)

## TODOs
Some work left in blocks.js (not the generator):
- [x] Update help urls
- [x] Update tooltips
- [x] Update hardcoded string constants

elatoskinas commented on appinventor/blocklyeditor/src/msg/en/_messages.js:
```diff
@@ -758,6 +758,14 @@ Blockly.Msg.en.switch_language_to_english = {
     Blockly.Msg.LANG_TEXT_TEXT_IS_STRING_TOOLTIP = 'Returns true if <code>thing</code> is a string.';
 
+    Blockly.Msg.LANG_TEXT_REPLACE_WITH_MAPPINGS_HELPURL = '/reference/blocks/text.html#replacemappings';
+    Blockly.Msg.LANG_TEXT_REPLACE_WITH_MAPPINGS_INPUT_TEXT = 'text';
```
TODO: Update data here

elatoskinas commented on appinventor/blocklyeditor/src/generators/yail/text.js:
```diff
@@ -318,3 +318,30 @@ Blockly.Yail['text_is_string'] = function() {
+    return [ code, Blockly.Yail.ORDER_ATOMIC ];
+  };
+
+  Blockly.Yail.text_replace_mappings.OPERATORS = {
```
@ewpatton Right now what I did was I took the order as a String parameter, and mapped it to an integer, which I pass in to the corresponding Java function as a parameter later on. Would this be a viable approach, or should we define a function for each mode in Scheme, and then call in to the appropriate function?

ewpatton replied:
I think it ultimately depends on how much code overlap there is. If it's saner to keep each as a different function you could map to three Scheme functions that call through to the Java implementation. If it's easier to pass an argument, then I would say pass the String itself and not bother mapping to an integer internally.

elatoskinas replied:
@ewpatton I figure keeping separate functions is a bit cleaner (since we do some form of mapping in the generator anyway). I did some refactoring; Let me know if you think it's overkill (I introduce a class for ordering in JavaStringUtils)

elatoskinas commented on appinventor/blocklyeditor/src/blocks/text.js:
```diff
@@ -497,3 +497,79 @@ Blockly.Blocks['text_is_string'] = {
+
+Blockly.Blocks.text_replace_mappings.TOOLTIPS = function () {
+  return {
+    LONGEST_STRING_FIRST : "tooltip",
```
TODO: Update tooltips

elatoskinas commented on appinventor/blocklyeditor/src/blocks/text.js:
```diff
@@ -497,3 +497,79 @@ Blockly.Blocks['text_is_string'] = {
+
+Blockly.Blocks.text_replace_mappings.HELPURLS = function () {
+  return {
+    LONGEST_STRING_FIRST : "help",
```
TODO: Update help urls

elatoskinas commented on appinventor/blocklyeditor/src/blocks/text.js:
```diff
@@ -497,3 +497,79 @@ Blockly.Blocks['text_is_string'] = {
+  },
+  //typeblock: [{translatedName: Blockly.Msg.LANG_TEXT_REPLACE_WITH_MAPPINGS_TITLE_REPLACE_WITH_MAPPINGS}]
+  typeblock: [{
+    translatedName: Blockly.Msg.LANG_TEXT_SPLIT_OPERATOR_SPLIT,
```
TODO: Update translatedName

elatoskinas commented on appinventor/blocklyeditor/src/generators/yail/text.js:
```diff
@@ -318,3 +318,30 @@ Blockly.Yail['text_is_string'] = function() {
 };
+
+Blockly.Yail['text_replace_mappings'] = function() {
+    // String replace with segment
```
TODO: Update description

ewpatton commented on appinventor/components/src/com/google/appinventor/components/common/YaVersion.java:
```diff
@@ -499,8 +499,10 @@ private YaVersion() {
   // For YOUNG_ANDROID_VERSION 201:
   // - CANVAS_COMPONENT_VERSION was incremented to 13
+  // For YOUNG_ANDROID_VERSION 202:
+  // - BLOCKS_LANGUAGE_VERSION was incremented to 30
```
I would recommend against bumping YOUNG_ANDROID_VERSION since there's no guarantee that this will be the next PR to be merged that affects YOUNG_ANDROID_VERSION. We will push a commit to your branch when we are ready to merge it to bump the version number here.

ewpatton commented on appinventor/components/src/com/google/appinventor/components/runtime/util/JavaReplaceWithMappings.java:
```diff
@@ -0,0 +1,189 @@
+// -*- mode: java; c-basic-offset: 2; -*-
```
I would consider merging the logic of JavaJoinListOfStrings and JavaReplaceWithMappings into a single utility class, e.g., JavaStringUtils. I think it's better than introducing a number of very small, targeted classes that hold only one or two methods each.

elatoskinas replied:
Agreed; Done!

ewpatton commented on appinventor/components/tests/com/google/appinventor/components/runtime/util/JavaReplaceWithMappingsTest.java:
```diff
@@ -0,0 +1,162 @@
+
+    assertEquals(expected, result);
+  }
+}
```
Missing newline at end of file.

ewpatton commented on appinventor/components/src/com/google/appinventor/components/runtime/util/JavaReplaceWithMappings.java:
```diff
@@ -0,0 +1,189 @@
+      Map.Entry<Object, Object> current = it.next();
+
+      // Get Key & Value, and update string mappings map
+      String key = current.getKey().toString();
```
We need to think about what to do if the map contains both the string "1" and the number 1.

elatoskinas replied:
@ewpatton The default solution in place (according to LinkedHashMap order) is to take the most recent (last) entry in the map (so effectively, new values override older ones). I think this is fairly reasonable, since we are replacing values in text, and there is no distinction in text between a number literal, and a number character. It might be worth adding in the description of the block or in the MD documentation that this would be the expected behavior.

elatoskinas replied:
Also might be worthwhile detecting this case explicitly, since both keys will be added to the List. Good point!

ewpatton commented on appinventor/components/src/com/google/appinventor/components/runtime/util/JavaReplaceWithMappings.java:
```diff
@@ -0,0 +1,189 @@
+      String key = keys.get(i);
+
+      // Append mapping that we want to replace to the regex pattern
+      patternBuilder.append(key);
```
Probably need to escape any `|` in key as `\\|`, otherwise they will be incorporated into the regex.

elatoskinas replied:
For the replacement itself, it should not affect the mapping itself, since we check for the key's presence before we apply the mapping when we encounter a regex match. However, that is a very fair point, since it can potentially make the regex slower (and not entirely correct). I will update to escape the keys.

ewpatton commented on appinventor/components/src/com/google/appinventor/components/runtime/util/JavaReplaceWithMappings.java:
```diff
@@ -0,0 +1,189 @@
+
+      // If there is still another mapping, then we append the union (OR) operator
+      if ((i+1) < keys.size()) {
+        patternBuilder.append("|");
```
Another approach to this is to have a `String sep = ""` outside the loop and then set it to `"|"` inside the loop. This way you never need to have the conditional.

elatoskinas replied:
@ewpatton I'm not sure I follow? The conditional is needed so that the union/OR operator does not get appended to the very last string, for example:
a | b | c

If we did not have the conditional, it would be:
a | b | c |

I am not really sure how this can be accomplished otherwise apart from looping backwards or changing the logic to another equivalent one (with more or less equal complexity)

ewpatton commented:
@elatoskinas I gave you some initial comments to go along with the TODO items. Please ping me when you think this is ready for a formal review.

elatoskinas commented on appinventor/components/src/com/google/appinventor/components/runtime/util/JavaStringUtils.java:
```diff
@@ -0,0 +1,240 @@
+  private static final boolean DEBUG = false;
+
+
+  // Implements the following operation
```
@ewpatton Should we move these comments elsewhere? Since the JavaJoinWithStrings class had some loose comments around, I moved them under the method. Let me know if you'd like to re-organize this a bit. Also, I moved the majority to the corresponding Javadoc.

elatoskinas commented on appinventor/blocklyeditor/src/blocks/text.js:
```diff
@@ -497,3 +497,79 @@ Blockly.Blocks['text_is_string'] = {
+Blockly.Blocks['text_replace_mappings'] = {
+  // Replace all occurrences in mappings with their corresponding replacement
+  category: 'Text',
+  helpUrl: Blockly.Msg.LANG_TEXT_REPLACE_WITH_MAPPINGS_HELPURL,
```
@ewpatton Is there any new functionality for auto-generating docs for blocks (like with the component markdown), or does it the same as before? (have to edit an .md/.html manually?)

ewpatton commented:
@stezelMIT I've deployed this at http://pr2085.ai2-ewpatton-temp.appspot.com if you want to try it out.

stezelMIT commented:
Oh wow!  Thank you! Thank you all!

elatoskinas commented on appinventor/components/src/com/google/appinventor/components/runtime/CloudDB.java:
```diff
@@ -1328,10 +1328,10 @@ private synchronized void ensureSslSockFactory() {
         tmf.init(keyStore);
         // // DEBUG
-        // Log.d(LOG_TAG, "And now for something completely different...");
+        // Log.d(LOG_TAG_JOIN_STRINGS, "And now for something completely different...");
```
Ah, seems like refactor all in my IDE bit back on me a bit, and I overlooked this. Thanks for pointing it out, I will revert it.

elatoskinas commented on appinventor/components/src/com/google/appinventor/components/runtime/util/AssetFetcher.java:
```diff
@@ -101,7 +101,7 @@ public static void upgradeCompanion(final String cookieValue, final String input
     //           form.startActivity(intent);
     //         } catch (Exception e) {
-    //           Log.e(LOG_TAG, "ERROR_UNABLE_TO_GET", e);
+    //           Log.e(LOG_TAG_JOIN_STRINGS, "ERROR_UNABLE_TO_GET", e);
```
Reverted

elatoskinas commented on appinventor/components/src/com/google/appinventor/components/runtime/TinyWebDB.java:
```diff
@@ -181,14 +181,14 @@ private void postStoreValue(String tag, Object valueToStore) {
     // developed, and then commenting them out when the component is
     // deployed.
-    // Log.w(LOG_TAG, "postStoreValue: sending tag = " +
+    // Log.w(LOG_TAG_JOIN_STRINGS, "postStoreValue: sending tag = " +
```
Reverted

elatoskinas commented:
@ewpatton That should cover the review changes above; We probably still want to add some documentation to the blocks to the website. Should this be done manually via the corresponding .md files?

ewpatton commented:
@elatoskinas Yes, you'll need to edit the markdown file to add new documentation. When you build the system (or run `ant docs`) it should update the HTML documentation.

elatoskinas commented:
> @elatoskinas Yes, you'll need to edit the markdown file to add new documentation. When you build the system (or run `ant docs`) it should update the HTML documentation.

@ewpatton Done; There is still the consideration left of how we should handle "1" and 1 cases if they are both present in the mappings (in one of the review comments left unresolved)

stezelMIT commented:
<img width="807" alt="Screen Shot 2020-03-12 at 12 53 34 PM" src="https://user-images.githubusercontent.com/44708504/76545756-b2f10780-6460-11ea-9da3-d32c5e52e5dc.png">
Thanks to @elatoskinas and everyone working on this.
I am running into errors of sorts.  Can someone explain to me why this returns "that me were good" whereas I would expect "that I was good" as "you were" is a longer string than "at you".
Also a question in general:  When we perform a replacement by longest string first, is the replacement made immediately before going to the next longest string replacement or are the replacements done at the very end?

elatoskinas commented:
@stezelMIT Ah, that is an interesting edge case there. The reason why the output is as you got is because the implementation I made as of now follows regex semantics, so in fact 'at you' is found first, hence it is also replaced first. The way the replacement happens is that we find a string by an union constructed by the keys in the specified order, and then we replace them as we go; So this all happens in a single pass (as it appears to us; but the regex matcher in use might be doing more work internally). So the replacement, in a sense, is made immediately.

So for the result you got, whether it's a bug or not depends on what semantics we view this as. If we put the regexp into a matcher, this is the result:
![image](https://user-images.githubusercontent.com/34982762/76571699-620cfd80-64b8-11ea-8816-868df279763e.png)

But I believe for our use cases it would make far more sense to actually replace the longest one first. I'll take a look at this.

stezelMIT commented:
Hi @elatoskinas,
Thank you again so much for your effort in this.
From the user's point of view when they select "longest string first" order (or the "dictionary" order) there is no ambiguity about what they need, so I am not sure how the following can be maintained as part of the replacement all algorithm:

  **"underlying order of matching is basically leftmost first, meaning that irregardless of the string length, we will always match by the key that appears first, and replace that first."**

I would assume first in the algorithm the strings would be sorted in the order requested by the user and then the reg exp procedure would be applied in that order.

@ewpatton if you have any wisdom to share (after your vacation is over, not before) kindly give us some guidance.
Thank you all

elatoskinas commented:
@stezelMIT I agree on the points you've mentioned. For the majority of use cases, it would make sense to replace in the order as specified in the block. However, it is also worth noting that this has the trade-off that now the order is completely altered, and we no longer have the ability to replace in the order of occurrence in the original string (which I believe has quite a few use cases as well, namely if the user wants to indeed apply a direct regex union operation as currently implemented). If we do consider the previously discussed "earliest occurrence" option, then the block becomes complicated in the sense that we again need 2 options for it ("longest string first" and "dictionary order").

Thus, for the sake of ambiguity, I would agree that perhaps changing it to the behavior you mentioned would make more sense. However, we should also take the trade-off into account, and perhaps it might be worth thinking about (any) implications caused by choosing one option over the other (here we will lose use cases in either way, unless we make a more complex block, but maybe one method is better than the other)

stezelMIT commented:
Dear @elatoskinas,
I believe the general sentiment here at App Inventor, as indicated by Hal's earlier comments, is to keep things simple as our main audience are K-12 students and teachers.  There are no plans at this point to implement "earliest occurrence" or other complex possibilities.  The replace all block just needs to do what it promises: make replacements according to either the "longest string first" preference or "dictionary order first" preference.  
@ewpatton Evan is due Wednesday and might be able to give more technical advice on how to make the implementations within the reg ex framework.  Thank you again so much for your interest in helping us out with this.

elatoskinas commented:
@stezelMIT Fair points indeed. I agree with you fully -- we should go to the suggested route to keep things simple, and implement a more simplified version (favoring the specified order) in the block. My initial confusion was in the sense that the pure regex implementation will do earliest occurrence first replacement by default (and actually, I don't think there's order redefinition), so there is some tweaking to do there to come to a more adapted solution to fit the needs. I will come back to a solution here soon.

@ewpatton 
For a more technical description, right now (internally) I'm thinking of the following:
1. Use regular expressions to find all matches, in the form of string index intervals [start, end]
2. Order the intervals by the specified ordering (dictionary/earliest occurrence first)
3. Replace the strings in the intervals by the mappings, preserving conflicts (if any character in interval is already replaced, we do not apply replacement)

Perhaps I'm overlooking into this and there is a more straightforward solution; But I feel like we need a manual adaptation here so as to preserve the replacements not being applied on top of each other.

stezelMIT commented:
1.  Order the intervals by the specified ordering (dictionary/earliest occurrence* first)
I think you mean * “longest string first” here.
Thank you.
Selim

From: Evaldas Latoškinas <notifications@github.com>
Reply-To: mit-cml/appinventor-sources <reply@reply.github.com>
Date: Monday, March 16, 2020 at 7:43 PM
To: mit-cml/appinventor-sources <appinventor-sources@noreply.github.com>
Cc: Selim Tezel <stezel@mit.edu>, Mention <mention@noreply.github.com>
Subject: Re: [mit-cml/appinventor-sources] Replace All Mappings block (#2085)


@stezelMIT<https://github.com/stezelMIT> Fair points indeed. I agree with you fully -- we should go to the suggested route to keep things simple, and implement a more simplified version (favoring the specified order) in the block. My initial confusion was in the sense that the pure regex implementation will do earliest occurrence first replacement by default (and actually, I don't think there's order redefinition), so there is some tweaking to do there to come to a more adapted solution to fit the needs. I will come back to a solution here soon.

@ewpatton<https://github.com/ewpatton>
For a more technical description, right now (internally) I'm thinking of the following:

  1.  Use regular expressions to find all matches, in the form of string index intervals [start, end]
  2.  Order the intervals by the specified ordering (dictionary/earliest occurrence first)
  3.  Replace the strings in the intervals by the mappings, preserving conflicts (if any character in interval is already replaced, we do not apply replacement)

Perhaps I'm overlooking into this and there is a more straightforward solution; But I feel like we need a manual adaptation here so as to preserve the replacements not being applied on top of each other.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/mit-cml/appinventor-sources/pull/2085#issuecomment-599806802>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AKVDFGCZFUJBKMCWBOOJPIDRH22SRANCNFSM4LDE2ASA>.

elatoskinas commented:
@stezelMIT Ah yes, you are right. I seem to be mixing up the terms in that regard!

I have implemented a new solution that no longer uses the standard regex order, but rather the redefined order to fit our needs. So now in the block you showed above, the result will be: "that I was good" indeed, since we first apply the "you were" mapping, and then no more mappings remain to be applied, since no more parts of text remain where the other mappings fit.

@ewpatton Interestingly enough, the change was quite complex. The idea that I did now is as follows:
**Issue:** The issue with regex order is that we cannot really redefine it. E.g. if I have the string `abcd`, and I have the union `bc|ab`, `ab` will be found first, no matter the order, because regex evaluates it from start to end (maybe there's a clever trick with lookaheads that I'm not aware of, but from what I could tell, regexp order cannot really be redefined due to the nature of it)

So the solution that I had in mind and implemented is to use, more or less, an interval tree data structure (I used `TreeSet` from Guava, since we already seem to have google commons in the codebase). The idea is as follows:
1. All the steps to constructing the post-processed map of mappings remains the same. We also re-order the keys in the same order (in the List). Then, we call in a function with our text, mappings and ordered key list to apply the mappings.

2. The first step is to process each key individually. We do this using the regex patterns & matchers, but now, instead of taking a union, we do it on a key-by-key basis (we process one key, then the next one, and so on). We do this because of the regex issue specified above.

3. In this key processing step, we can use the matcher to find the next occurrence. From this occurrence, we can extract the start & end indices. We construct a `Range` object from these indices (to use in our `RangeSet`), and we check whether the range overlaps with all the intervals in the `RangeSet` (if the RangeSet is implemented as per the Interval tree data structure, this should be `O(log n)` time. Provided the range does not overlap, we add it to the `RangeSet`

4.In addition to the above, we will keep a Map (let's say `ReplacementMap` from `Range` to `String`. This essentially tells us what range of the string we want to replace with which `String` for ease of use. This map will be ordered in **descending** order of the range endpoint (it will be a `TreeMap`)

5.After processing all the keys, we iterate the `ReplacementMap`'s entries, and apply the replacement in order. We use the combination: ` L + M + R `, where ` L ` is the string of text up until the current range's lower bound index, `M` is the replacement string (the value of the mapping) and `R` is the remainder of the string after `M`.

There is one key observation to note, specifically the map's chosen order in step 4. If we do ordering in a different way, we will run into issues that the ranges will no longer be accurate (we will need to update their indices). For example, consider:
Ranges: `[0 4], [5, 6]`. Suppose we replace `[0, 4]` with a substring of size 7. Well, now all the other ranges have shifted by `3`, and this becomes harder to keep track of.
However, if we replaced starting from `[5,6]`, the indices of `[0,4]` do not change, because we only replaced the right end of the string.
And because the ranges don't overlap, this ordering should be correct.

That was quite lengthy, but this solution seemed to work for the test cases that we've had so far. I've tried to document this as thoroughly as possible in the current code.

ewpatton commented:
@stezelMIT I've pushed the new version to the test server.

@elatoskinas I'm not able to compile apps with this version. Companion works okay though.

elatoskinas commented:
> @elatoskinas I'm not able to compile apps with this version. Companion works okay though.
@ewpatton This  seems to be an extremely weird problem, one that might seem to cause some indirect crashes.. I identified the crash to be caused in the `applyMappings` method that does all the heavy-lifting, as commenting out did not crash the app building.

However, I cannot seem to understand the following phenomenon. This does not crash:
```java
    // Create a set of ranges to keep track of which index ranges in the
    // original text string are already set for replacement.
    RangeSet<Integer> ranges = TreeRangeSet.create();

    // Map to map Range to String to replace with.
    // E.g. [1, 3) -> 'abcd' indicates that the substring from 1 to 3 exclusive should
    // be replaced with the string 'ab'.
    // Note that the length of the string does not matter.
    Map<Range<Integer>, String> replacements = new TreeMap<>(new Comparator<Range<Integer>>() {
      @Override
      public int compare(Range<Integer> r1, Range<Integer> r2) {
        // Sort in descending order
        return Integer.compare(r2.upperEndpoint(), r1.upperEndpoint());
      }
    });

    // Range construction step: Iterate through all the keys,
    // and fill in ranges set & replacements map.
//    for (String key : keys) {
//      // omitted commented out code
//    }
 
     // more commented out code here ommited

    return text;
```

Yet, this crashes:
```java
    // Create a set of ranges to keep track of which index ranges in the
    // original text string are already set for replacement.
    RangeSet<Integer> ranges = TreeRangeSet.create();

    // Map to map Range to String to replace with.
    // E.g. [1, 3) -> 'abcd' indicates that the substring from 1 to 3 exclusive should
    // be replaced with the string 'ab'.
    // Note that the length of the string does not matter.
    Map<Range<Integer>, String> replacements = new TreeMap<>(new Comparator<Range<Integer>>() {
      @Override
      public int compare(Range<Integer> r1, Range<Integer> r2) {
        // Sort in descending order
        return Integer.compare(r2.upperEndpoint(), r1.upperEndpoint());
      }
    });

    // Range construction step: Iterate through all the keys,
    // and fill in ranges set & replacements map.
    for (String key : keys) {
      // omitted commented out code
    }
 
     // more commented out code here ommited

    return text;
```

The only difference is that I uncommented out the for-loop there, which has an empty body. Is this some strange memory issue that we are running into? (although, would assigning one String of memory really break it in such a way?)

I will try to look at it tomorrow again with fresh eyes, but this seems very strange.

ewpatton commented:
@elatoskinas It's not even to that point. I mean that when I run the dev server and build server, even with a blank project I get an error in the Kawa stage at compiling.

ewpatton commented on appinventor/components/tests/com/google/appinventor/components/runtime/util/JavaStringUtilsTest.java:
```diff
@@ -172,9 +173,102 @@ public void testEscapeUnionCharacter() {
+   * String with a mapping that maps an empty string to some String
+   * replaces all gaps in the String.
+   *
+   * TODO: Is this expected behavior?
```
This is expected behavior as at each index in the string the empty string trivially matches.

ewpatton commented on appinventor/components/src/com/google/appinventor/components/runtime/util/JavaStringUtils.java:
```diff
@@ -0,0 +1,431 @@
+   * @param mappings String -> String mappings to replace from Key to Value
+   * @return String with mappings applied
+   */
+  /*private static String applyMappings(Matcher matcher, Map<String, String> mappings) {
```
Please remove unused code if it is no longer necessary.

ewpatton commented on appinventor/components/src/com/google/appinventor/components/runtime/util/JavaStringUtils.java:
```diff
@@ -0,0 +1,431 @@
+   * @param keys  List of keys to union together
+   * @return  Single Regex String of keys combined by unions
+   */
+  /*private static String buildUnionFromStrings(List<String> keys) {
```
Ditto

elatoskinas commented:
> @elatoskinas It's not even to that point. I mean that when I run the dev server and build server, even with a blank project I get an error in the Kawa stage at compiling.

@ewpatton Yes; that's correct. But what I meant is, commenting the code as above will no longer make it crash at that stage. Meanwhile, uncommenting just one for-loop with a blank body will crash prior to the compilation stage, which is very odd.

ewpatton commented:
@elatoskinas Let me see if I can repro that on my end.

elatoskinas commented:
@ewpatton I found the underlying cause. When I did the following instantiation:
`RangeSet<Integer> ranges = TreeRangeSet.create()`

`RangeSet` being a superclass of `TreeRangeSet`, for some reason this made the apk builder act up and crash. Weirdly enough, replacing with:
`TreeRangeSet<Integer> ranges = TreeRangeSet.create()`

Solved the problem.

This is a very weird subtle bug.. At this point I'm really confused whether this is a Kawa, Guava or a build issue in AI locally

ewpatton commented:
@elatoskinas Thanks for the sleuthing. Confirmed that your most recent change fixes the issue. Can you go ahead and merge ucr to resolve the conflicts?

elatoskinas commented:
> @elatoskinas Thanks for the sleuthing. Confirmed that your most recent change fixes the issue. Can you go ahead and merge ucr to resolve the conflicts?

Done. However, we still need to resolve the EDIT in the comment above, which occurs when using the new block in compiled .apks:
![image](https://user-images.githubusercontent.com/34982762/77194213-d27ed480-6adf-11ea-926c-5c9057b274c7.png)

I see that Guava is in the build.xml for components, but somewhat the Guava class does not get resolved. Is this somehow related to the way runtime.scm calls into Java methods, and do we also need to update a build.xml in that regard?

ewpatton commented:
@elatoskinas Ah, right. The guava library is conditionally included for certain components. It's not included in all builds, which is probably why builds for simple apps failed but it worked for the companion. You may be able to copy the TreeRangeSet source into the com.google.appinventor.components.runtime.collect package. I don't know how tightly coupled it is to the rest of the Guava system though.

elatoskinas commented:
@ewpatton It seems to be quite coupled in this regard (at least ~9 classes, although some of them are annotations, but the major problem is the classes that are subclassed from), so copy pasting it directly is not as straight forward it seems.

Do we have other alternatives before we resort to this? Also dropping Guava to implement an interval tree is an option (although will end up being tedious, complex and hard to maintain), or we might look for other solutions if this doesn't work out.

ewpatton commented:
I think it wouldn't be that difficult to implement the logic with TreeSet and a custom comparator that sorts the ranges by start index.

elatoskinas commented:
> I think it wouldn't be that difficult to implement the logic with TreeSet and a custom comparator that sorts the ranges by start index.

@ewpatton It depends. If we are aiming for a naive solution that does the job, then yes, it's probably not too difficult (although could lead to range re-combination solutions that would ultimately lead to beyond quadratic, cubic possibly time in the usage of the block).

If we want the solution to be a bit more scalable (although, understandably, that typically is not a concern for the majority of the use cases), the interval tree is quite complex to implement from scracth, in the sense that it will be hard to maintain (e.g. the [TreeRangeSet](https://github.com/google/guava/blob/master/guava/src/com/google/common/collect/TreeRangeSet.java) in Guava has 44 lines of code roughly for the add method alone. When we add everything we need, we have 200-300 more for a supplementary data structure alone).

We could always go with the more naive approach, but if possible, I am always up for re-using solutions that already exist, are stable and fully tested. I will try to look through the Guava class again and see if maybe we can somehow extract only the needed bits while dropping some functionality, provided that conditionally including this class straight form Guava is not an option

ewpatton commented:
One problem is that we don't have a way of conditionally including a library for a specific block. Another is that Guava is pretty big to provide support for just a single block. Unless I'm mistaken, it looks like you only need the TreeRangeSet to test whether there is overlap between the current match and existing matches. This should be really quite doable within the existing Java constructs like TreeSet and its ability to return iterators to elements before/after a given value.

elatoskinas commented:
@ewpatton Agreed on Guava; If conditional inclusion is not an option, then it is indeed to big for our sake. Also, you had a very good point with TreeSet; For some reason I read TreeRangeSet initially and thought you meant re-implementing the logic posed there somehow. But using overlap checks in the comparator, the TreeSet indeed suffices. I have pushed the rewritten changes to no longer use Guava.

ewpatton commented on appinventor/components/src/com/google/appinventor/components/runtime/util/JavaStringUtils.java:
```diff
@@ -290,18 +310,28 @@ public static String replaceAllMappings(String text, Map<Object, Object> mapping
+    // original text string are already set for replacement, together with the
+    // text to replace with. We sort this TreeSet in descending order to preserve
+    // indices of all ranges after replacement.
+    TreeSet<Range> ranges = new TreeSet<Range>(new Comparator<Range>() {
```
Minor nit: You could make this `static final` on the class so we only instantiate one.

elatoskinas replied:
You mean the comparator, right?

ewpatton replied:
Yes, the comparator only needs to be constructed once.

elatoskinas replied:
Done!

ewpatton commented:
I tested this in both companion and compiled apps. The approach seems to work as expected.

ewpatton commented:
Just waiting on the change w.r.t. the comparator and then I think we can merge it.

ewpatton commented on appinventor/components/src/com/google/appinventor/components/runtime/util/JavaStringUtils.java:
```diff
@@ -140,11 +165,13 @@ public Range(int start, int end, String text) {
+  private static final Comparator<Range> rangeComparator;
 
   static {
     mappingOrderDictionary = new MappingOrder();
```
Is there any reason you need these to be initialized in a `static` block rather than just assigning them when the variable is declared?

elatoskinas replied:
Good point; There's no reason to have it anymore, before that I used it when the code was structured in such a way that required adding things to a data structure. I've removed it.
